hal/tls/mbedtls/tls_mbedtls.c

```diff
@@ -920,22 +920,30 @@ TLSSocket_read(TLSSocket self, uint8_t* buf, int size)
         return -1;
     }
 
-    int ret = mbedtls_ssl_read(&(self->ssl), buf, size);
-
-    if ((ret == MBEDTLS_ERR_SSL_WANT_READ) || (ret == MBEDTLS_ERR_SSL_WANT_WRITE))
-        return 0;
+    int len = 0;
+    while (len < size) {
```

Comment:
I'm not sure why you put the loop here. I get the idea: you will loop reconstructing the packet until the desired size is reached.

But this as far as I understand is already done at a higher level in `CotpConnection_readToTpktBuffer` where a read is done (call to this function), then a check if we have enough needed data and if not it will return `TPKT_WAITING`  which in `IsoClientConnection_handleConnection` will signal that the packet is not complete and will then retry later (using `MmsConnection_tick`) which also manages a kind of thread relaxation (via `Thread_sleep` calls).

So I'm not sure this change of logic is a good idea (you made the read waitiling loop at a lower level compared to before which was at state machine level), I would just in case augment the `ret` comparison with the two new `MBEDTS_ERR_SSL_` and that's all.

Reply:
Mmm, you're right and I had your same thoughts and doubts.  
Here are my supposition, so take it with a big grain of salt:

Handling at a higher level, with sleeps and so on, left out the proper handling of `MBEDTLS_ERR_SSL_ASYNC_IN_PROGRESS` and/or `MBEDTLS_ERR_SSL_CRYPTO_IN_PROGRESS` which I think is happening *during* a renegotiation.
This not properly read buffer (or delayed read) left the door open to an incoplete data read which resulted renegotiations always ending with `Handshake failed for unknown reason`.

Also I suspect that, in a similar way of the session reset, the reads may end up being "interleaved" with other operations which then are using a "dirty" read/write buffer.

Again, please take this as is: just ideas! As I found it to very hard for me to follow the flow logic of the events, also because this is happening with INTENSE reads/writes.

I can eventually give it some more tests in a later moment.